renku/core/dataset/providers/api.py

```diff
@@ -177,39 +177,52 @@ def update_files(
             communication.start_progress(progress_text, len(files))
 
             storage = self.get_storage()
-            hashes = storage.get_hashes(uri=files[0].dataset.storage)
+
+            # group files by storage to efficiently compute hashes
+            storage_files_dict: Dict[str, List["DynamicProxy"]] = defaultdict(list)
+
             for file in files:
-                communication.update_progress(progress_text, 1)
-                if not file.based_on:
-                    continue
-
-                dst = project_context.metadata_path.parent / file.entity.path
-
-                hash = next((h for h in hashes if h.uri == file.based_on.url), None)
-
-                if hash:
-                    if not dry_run and (
-                        not file.dataset.storage
-                        or not is_uri_subfolder(resolve_uri(file.dataset.storage), file.based_on.url)
-                    ):
-                        # Redownload downloaded (not mounted) file
-                        download_storage = self.get_storage()
-                        download_storage.download(file.based_on.url, dst)
-                    file.based_on = RemoteEntity(checksum=hash.hash if hash.hash else "", url=hash.uri, path=hash.path)
-                    results.append(DatasetUpdateMetadata(entity=file, action=DatasetUpdateAction.UPDATE))
+                if file.dataset.storage:
+                    storage_files_dict[file.dataset.storage].append(file)
                 else:
```

Comment:
I don't think the above is true. if I just do `renku dataset add --create s3://bucket/myfile.csv`, it will get a `based_on` set to a remote entity. As in adding a single s3 file to a non-storage backed dataset.
That's what the above code is handling.

The whole `update_files` method on the provider should only be called with files relevant to that provider anyways, but I can add a sanity-check/exception if for instance S3Provider.update_files is called with a non-s3 file